### ISSUE TITLE
Proposed RFC Suggestion - O3DE Code Review Guidelines

### ISSUE THREAD

**Issue by hultonha** [o3de/community#140]
> Note: Please scroll to the end of the document to see an early form of the proposed document

## Summary

A set of guidelines for code reviewers and change authors when contributing to Open 3D Engine (O3DE). This is to help pull requests (PRs) go more smoothly and help the community share best practices/lessons learned when it comes to changes and reviews.

### Why is this important? 

Publishing PRs (especially for new contributors) is a scary process and giving and receiving feedback can also be very difficult to do well (even for experienced engineers). We would like to draft a set of guidelines to help both new and existing contributors create a supportive and inclusive environment around PRs to maximize the efficiency and efficacy of changes made to O3DE.

### What are the use cases for this suggestion?

A short, succinct doc that new (and existing) contributors can review to understand what makes a good PR and what makes a good review. What are some of the things you should do, and what things are best to avoid. This document can give positive examples and be used as a resource or refresher to return to if clarification is ever needed on the PR process.

### What should the outcome be if this suggestion is implemented?

A new wiki document hosted on the O3DE Community repo called 'Code Review Guidelines' that we can share liberally with new and existing contributors to improve the entire communities' experience of PRs.

### Example wiki markdown document

```
# O3DE Code Review Guidelines

## Useful Related Material

- [O3DE C++ Coding Standards](https://github.com/o3de/sig-core/blob/main/governance/Coding-Standards-and-Style-Guide.md#o3de-c-coding-standards-)
- [O3DE API Ref Guidelines](https://github.com/o3de/sig-core/blob/main/governance/API-Ref-Guidelines-Update.md)

## Change Author Guidelines

... details to follow about best practices/advice for change authors

## Code Reviewer Guidelines

... details to follow about best practices/advice for code reviewers

```

#### New terminology

- Change author, the person publishing the PR
- Code reviewer, the person reviewing the PR

### What are the advantages of the suggestion?

- Help new contributors onboard to O3DE and build a more welcoming, inclusive and transparent environment.
- Make sure PRs have the greatest chance of success and contributions to O3DE are successful.

### What are the disadvantages of the suggestion?

- The reality is this is yet one more document to read, but as this is such a critical aspect of open source development, the benefits of having it outweigh the costs of relying on tribal knowledge and good intentions.

### How will this be work within the O3DE project?

- This document will be recommended reading for all new and existing contributors to O3DE to help people all get on the same page about how to be an effective change author and code reviewer.

### Are there any alternatives to this suggestion?

- Do nothing. Rely on tribal knowledge and other community engagement (Discord, PR comments, GitHub Discussions etc...) 
- There is a risk new contributors will be discouraged from contributing without a framework to help them make changes and also get feedback and comments in reviews that are constructive and helpful.
- Rely on existing public code review guidance (a great example from Google can be found [here](https://google.github.io/eng-practices))

### What is the strategy for adoption?

- Add this to the contributor guidelines doc (sigs can also recommend reading this in their charters)
- Messaging on Discord and through SIG meetings can help raise awareness and drive engagement

---

> Begin actual document

# O3DE Code Review Guidelines

This document contains a set of guidelines for code reviewers and change authors when contributing to Open 3D Engine (O3DE). The intention is to help pull requests (PRs) go smoothly and help the community share best practices/lessons learned when it comes to changes and reviews.

## Useful Related Material

- [O3DE C++ Coding Standards](https://github.com/o3de/sig-core/blob/main/governance/Coding-Standards-and-Style-Guide.md#o3de-c-coding-standards-)
- [O3DE API Ref Guidelines](https://github.com/o3de/sig-core/blob/main/governance/API-Ref-Guidelines-Update.md)

### Keep it short

This of course isn't always possible, there will be exceptions, but to get a PR approved and merged in a reasonable time-frame, try to keep changes as contained and atomic as possible. Small incremental progress is much safer and simpler than a single large change that contains a many modifed lines and files.

### Review your own PR

The best way to catch simple mistakes and small oversights is to read through your own PR in the GitHub UI before hitting the publish button. Viewing your own code through this slightly different lens often uncovers a lot more issues than you'd expect and can help save other reviewers time.

### PRs are complete

Try not to publish something for review when still making changes. This can lead to reviewers needing to re-review changes or confusion in what has changed or not. Try to only make requested changes in a PR to avoid unnecessary churn.

**Draft PRs**: _Note: This does not apply to 'Draft' PRs which are great to use to share early ideas/designs. The reviewer knows they aren't yet 'reviewing' the code, they're just giving initial feedback. Keeping a PR in a draft state also ensures it cannot accidentally be merged._ 

### Acknowledge/respond to comments

It's good practice to acknowledge you've read the reviewer's feedback. This can be as simple as leaving a 👍 or a brief comment and then resolving the conversation. If the comment was simple and has been fixed, it's usually fine to resolve the conversation. Try to avoid resolving conversations without acknowledging or responding to them either with a reply or emoji.

### Don't sweat the small stuff

Reviewers will sometimes leave optional/suggested feedback. Not everything has to be accepted, especially with certain stylistic choices (the Coding Standards should be the ultimate arbiter here to help resolve any conflicts). It's often appropriate to move some feedback to a follow-up PR that makes the relevant updates, or they can be rolled into a future change.

### Avoid mixing functional and formatting changes

When making updates to the code, try to avoid making numerous formatting only changes while also including functional updates. The best approach is to make the functional change with the most minimal, necessary changes, and create a separate PR that exclusively includes the formatting changes.

### A picture is worth a thousand words

If your change results in any kind of visual difference, to help reviewers quickly get context on the PR, favor including screenshots and/or videos/gifs of the change in action (before/after is also very useful). This helps frame the change and can make reviewers lives much easier to see what the effect of the change is.

### Engage UI/UX when needed

If your PR involves user interface (UI) or user experience (UX) changes, please ensure you have read how to get help from [SIG/UI-UX](https://github.com/o3de/sig-ui-ux/blob/main/governance/UX%20Intake%20Process.md). SIG maintainers can help you connect and engage with the UI/UX team.

### Engage Docs when needed

If your PR requires changes to the O3DE documentation or tutorials, ensure you've reviewed https://github.com/o3de/sig-docs-community/issues/43 [currently in review]. SIG maintainers can help you connect and engage with the documentation experts.

### Nudges are okay

If your PR has been waiting a while and no one has looked at it, it's okay to start using gentle nudges to ask for feedback. This most likely will be on Discord but it's possible to also tag people in the PR comments and politely ask them to take a look. Code owners will be automatically notified but sometimes things can be missed.

### Be kind

> (Stolen from Google's excellent [engineering practices](https://google.github.io/eng-practices/review/reviewer/comments.html))

Writing code is hard and doing so in a new environment with unfamiliar faces can be even more daunting. As reviewers we want to encourage and support new contributors. We want to thank people for their efforts and offer constructive feedback where necessary. Remember to call out the highlights of a change as well as things that need improvement.

### It's about the code

When writing review feedback focus on what the code is doing, not what the change author did. Avoid 'you' if at all possible. Instead use 'we' to demonstrate our shared responsibility and ownership of the code base or refer to the code directly.

e.g.
Avoid: *"You should use a vector here has it will lead to better cache locality"*
Prefer: *"A vector might be a good idea to use here to get better cache locality"*

### Include the _why_

When suggesting a change (unless it's something really trivial) it's helpful to include the reason _why_ such a change is being proposed. This gives context to the request and can often act as a learning opportunity for the code author and other contributors who may be reviewing or observing the PR.

### Don't be late to the party

If a PR already has two approvals (from at least one code owner) be conscious that additional feedback will interrupt the flow of a PR. Feel free to review and if a serious issue is spotted do feel empowered to take the necessary steps (request changes) but try to avoid lots of additional comments that could be addressed in a follow-up PR.

### Keep feedback focused

Ensure feedback is targeted at the changed lines of code and not surrounding code. It isn't the changes authors responsibility to fix a problem a reviewer sees that happens to be in that area of code. If something is spotted, make an issue, and follow-up with a separate PR.

### Be judicious with 'Request Changes'

Request changes is a powerful tool and shouldn't be used lightly. If an apparent bug, performance problem or architecture issue is spotted, then 'Request Changes' may be used. However, often simply not approving the PR and leaving a comment is sufficient to resolve whatever problem is detected, and does so in a softer and friendlier way. Prefer this option if you can.

### Be responsive

After reviewing a PR, keep an eye on when the change author has responded and do your best to reply to their responses promptly (ideally within a 24 hour window, though this isn't always possible). If you are delayed and don't have time to respond fully, leave a comment on the PR that you're busy and will get back to it as soon as you can.

### Approve if you can

> Note: This sentiment is also lifted from the excellent Google [engineering practices](https://google.github.io/eng-practices/review/reviewer/standard.html) page

_In general, reviewers should favor approving a [PR] once it is in a state where it definitely improves the overall code health of the system being worked on, even if the [PR] isn’t perfect._

It's inevitable all changes will not get everything right, but if what is there improves the software while keeping it maintainable and efficient, a reviewer should feel empowered to approve.

### If you can't approve, take things offline

If there is a point of discussion on a PR that cannot be quickly resolved, consider reaching out directly to the change author to talk it over (either over Discord, a separate GitHub discussion or even an RFC - sometimes a voice call is even better if possible). In our experience this leads to better outcomes for both parties and avoids potential misunderstandings. Leaving a brief update on the PR with the resolution afterwards is good practice too.

## Contributors, Reviewers and Maintainers

Once a PR is approved by two or more reviewers, an AR run can commence. Currently only [maintainers](https://github.com/orgs/o3de/teams/maintainers) can initiate builds so please feel free to ask for someone with access to start a build. The list of PRs can be found here - [Change Requests](https://jenkins.build.o3de.org/job/O3DE/view/change-requests/). Search for the PR number to review the current status of the build. Once AR has completed and there are no failures, a PR can be merged by a maintainer.

**Comment by lmbr-pip** [comment]
Personally, I don't see how anyone would object to having guidelines, I can see folks objecting to the specific guidelines. Have we considered providing draft of initial guidelines in this RFC? For example, would be really useful to link to our great but very hard to find coding standards ;)

Also RFC, as its the meta RFC, should propose specifically where they live (assume community, but it may be good to include repro and location). We may also want to consider updating the PR request template to include a link to guidelines, so PR authors can pre-flight changes against guidelines.

**Comment by lumberyard-employee-dm** [comment]
I think having a guidelines document for Pull Requests are fine to add to the PR request template.

Specifically @hultonha what issues are you looking to address with the guidelines RFC.
The advantages mention providing a welcoming and inclusive environment for PRs, which is good.

For the second advantage I wonder what would increase the chance of PR success.
Is it a good description by the change author, correct use of the Request Changes vs Comments by the code reviewer,etc...

**Comment by hultonha** [comment]
@lmbr-pip That's a good call, I can include a section/line break at the end of the RFC above with the proposed document that we can discuss in more detail (instead of the outline currently in the code section). Also on a side note, there is a plan to move the coding standards and API guidelines to the community repo as well so everything (including this and deprecation docs) are all in the same place. I'm talking with @aFinchy about this.

Yes this would be the community repo, but I can make this more clear in the RFC itself. I think linking to it after is something we can do separately once it exists if we think it'd be a good idea. Thanks for the feedback!

@lumberyard-employee-dm Great to hear you're supportive of the idea, thanks for your feedback. The main issues we'd like to address are 'what makes a good change' (cover things like, don't change too much if you can help it, don't change many unrelated things, avoid mixing formatting and functional changes, include a helpful description, be responsive/acknowledge PR feedback etc...), and also what makes a good review (when to use 'request changes', how to make suggestions, what to focus on, how to be supportive and positive when giving critical feedback etc...).

I think the set of guidance for change authors should improve the chance of PR success and the guidance for reviewers should hopefully improve the experience for change authors too and make it more likely they'll return for future PRs. Thanks!

**Comment by hultonha** [comment]
Hey there - I've left a significant update to the doc with the proposed document content below the RFC proposal. Feel free to offer suggestions here or we can hopefully soon move this to an actual PR and discuss the points there. Thanks!

**Comment by lmbr-pip** [comment]
This looks great, thanks for doing this.

Add link to the UI/UX process? as that can impact a lot of PRs.

**Comment by hultonha** [comment]
Thanks @lmbr-pip! Ah that is a good shout, I don't suppose you have a link to hand? 😛 Otherwise I'll ping the sig-ux folks. Cheers!

**Comment by lmbr-pip** [comment]
https://github.com/o3de/sig-ui-ux/blob/main/governance/UX%20Intake%20Process.md

**Comment by hultonha** [comment]
Great suggestion, I can add this now 👍

**Comment by hultonha** [comment]
> Something like?
> ### Engage the user-experience experts when needed
> 
> If your PR involves user-interface (UI) changes, ensure you have read how to [get help from SIG/ui-ux](https://github.com/o3de/sig-ui-ux/blob/main/governance/UX%20Intake%20Process.md). SIG maintainers can help you connect and engage with the UI-UX experts.

Added and also added an entry about including a screenshot/video/gif of the change to help both UI/UX and reviewers.

**Comment by lumberyard-employee-dm** [comment]
I actually have a bit tweaks to the `Be judicious with 'Request Changes'` section.

> Request changes is a powerful tool and shouldn't be used lightly. If a serious bug, performance problem or architecture issue is spotted, then 'Request Changes' may be used. However, often simply not approving the PR and leaving a comment is sufficient to resolve whatever problem is detected, and does so in a softer and friendlier way. Prefer this option if you can.

I think if author is adding an obvious bug, if it is not serious it should be marked with requested changes.
I don't think code should be accepted with obvious bugs when the fix is not difficult to make.

Now on separately, my opinion on the "Request Changes" option, is that it is a required change, while all regular comments are optional to address, but that is difference in reviewing styles, that we need to come to a consensus to.

Next around the `It's about the code` section when giving feedback, it is important to emphasize that the reviewer explains why a change is needed.

The example that is there already does a good job of explaining why the change is needed, but we want to put a blurb in about it.
How about the following amendment to that section

> When writing review feedback focus on what the code is doing, not what the change author did. Avoid 'you' if at all possible. Instead use 'we' to demonstrate our shared responsibility and ownership of the code base or refer to the code directly.
_Try to explain to the change author why the change is needed and how it improves the code quality._


For the `PRs are complete` section, the blockquote text around Draft PRs is drawing attention away from that section, de-emphasizing it stylistically. Perhaps we can make smaller header for that part, so it doesn't appear grayed out.
Additionally, we may to mention that Draft PRs should be used instead of having the a description start `[DO NOT MERGE]`.
A Draft PR isn't can't be merged by mistake, while an Open PR with the `[DO NOT MERGE]` description can be mistakenly merged by a reviewer or maintainer.


Next, I think this document should mention how change authors can access the build results through Autormated Review(AR) of there PRs, as well as to a section mentioning that reviewers with "Maintainer" and "Reviewer" privileges should be cognizant of when a PRs is ready to build on Jenkins and to kick off those builds.
External Contributors don't have access to kicking off builds, so a bit of time can be saved by recognizing that the PRs hasn't had an AR started yet and doing so on the author behalf.

**Comment by hultonha** [comment]
Thank you for the feedback @lumberyard-employee-dm, much appreciated. I'm happy to make some edits now but ideally just want to get the RFC approved so I can create a PR with the doc and we can go back and forth a little on exact working there.

To your points:

> I think if author is adding an obvious bug, if it is not serious it should be marked with requested changes. I don't think code should be accepted with obvious bugs when the fix is not difficult to make.

This is totally fair, I can update the wording to say 'obvious' or maybe 'clear', though I do think the severity should play a part. For example maybe passing by value for a string type (when it's not a 'sink' parameter) could be classed as a bug, but if it's not in performance critical code and not actually breaking anything, adding 'request changes' seems too much, a comment with a clear indication we should update this feels sufficient and is not overly punitive, so again it is up to reviewers to use their best judgement.

> Now on separately, my opinion on the "Request Changes" option, is that it is a required change, while all regular comments are optional to address, but that is difference in reviewing styles, that we need to come to a consensus to.

I think specifying 'update' in the comment for a change (without the massive, red, **Change Request** label should be the default (see guidance above). Hopefully we can move people towards this as I think it'll make reviews overall less adversarial and more positive.

> Draft PRs is drawing attention away from that section, de-emphasizing it stylistically. Perhaps we can make smaller header for that part, so it doesn't appear grayed out.

Good call, this is maybe the fault of Markdown or GitHubs styling choices, I can reformat this slightly to make it standout more.

> Next around the It's about the code section when giving feedback, it is important to emphasize that the reviewer explains why a change is needed.

Great point about including the _why_, I think include a section for this on its own. It helps give context and helps developers/contributors learn (even those not directly participating in the review).

> Next, I think this document should mention how change authors can access the build results through Autormated Review(AR) of there PRs, as well as to a section mentioning that reviewers with "Maintainer" and "Reviewer" privileges should be cognizant of when a PRs is ready to build on Jenkins and to kick off those builds.

Also a great point, I hadn't focused so much on these more boilerplate parts but they are super important. I'll add a section to cover this.

Thanks for the feedback 👍

**Comment by hultonha** [comment]
Updates have been made, let me know what you think (also we can iterate in the PR too once the RFC is approved). Thanks!

**Comment by lgleim** [comment]
Similar to involving SIG/ui-ui, there should probably also be a section on how to involve sig/docs, especially for any PR that might change or extend APIs.

### Engage SIG/docs when needed

If your PR requires changes to the O3DE documentation or tutorials, ensure you have read how to [get help from SIG/docs](https://github.com/o3de/sig-docs-community/issues/43) [currently in review]. SIG maintainers can help you connect and engage with the documentation experts.

**Comment by hultonha** [comment]
Thanks for the suggestion @lgleim, happy to add this but we can totally get this perfect during the PR. I'm just looking for approval of the general concept (RFC) that having review guidelines is a good idea. We can then formulate them in the actual PR for adding to this repo. Thanks!

**Comment by Ulrick28** [comment]
Definitely agree with this proposal and thanks for taking this on! I agree with @lmbr-pip that I would be surprised if anyone felt this was a bad idea. At this point, I believe you have enough buy-in to start on the RFC

**Comment by hultonha** [comment]
Thanks for the feedback @Ulrick28, much appreciated 🙂Excellent, to make it official would you please be able to close the issue and I'll create a follow-up PR that references this as soon as that's done. Thanks!

**Comment by hultonha** [comment]
(also added your suggestion @lgleim, thanks!)

**Comment by Ulrick28** [comment]
Closed as requested

**Comment by hultonha** [comment]
Thanks @Ulrick28! 🙂
